File: src/wpsr/parsers/dom/DOMCandidateManager.java
package wpsr.parsers.dom;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import org.apache.log4j.Logger;
import wpsr.factory.PredicateFactory;
import wpsr.parsers.rdf.predicates.Predicate;
import wpsr.parsers.rdf.query.IRDFNode;
import wpsr.parsers.rdf.query.IRDFStatement;

public class DOMCandidateManager {

    static Logger logger = Logger.getLogger(DOMCandidateManager.class);
    private PredicateFactory predicateFactory = new PredicateFactory();
    private DOMProcessor processor;
    private HashMap<String, ArrayList<IDOMCandidate>> scoreboard;
    private HashMap<String, ArrayList<IDOMCandidate>> exactMatch;

    public DOMCandidateManager(DOMProcessor processor) {
        exactMatch = new HashMap<String, ArrayList<IDOMCandidate>>();
        scoreboard = new HashMap<String, ArrayList<IDOMCandidate>>();
        this.processor = processor;
    }

    public void exactMatch(String uri, IDOMCandidate candidate) {
        if (candidate == null) {
            return;
        }
        if (exactMatch.get(uri) == null) {
            final ArrayList<IDOMCandidate> cContainer = new ArrayList<IDOMCandidate>();
            cContainer.add(candidate);
            exactMatch.put(uri, cContainer);
        } else {
            final ArrayList<IDOMCandidate> cContainer = exactMatch.get(uri);
            if (cContainer != null) {
                cContainer.add(candidate);
                exactMatch.put(uri, cContainer);
            }
        }
    }

    public void remove(String uri, IDOMCandidate candidate) {
        if (candidate == null) {
            return;
        }
        if (scoreboard.get(uri) == null) {
            return;
        } else {
            scoreboard.get(uri).remove(candidate);
        }
    }

    public void add(String uri, IDOMCandidate candidate) {
        if (candidate == null) {
            return;
        }
        if (scoreboard.get(uri) == null) {
            final ArrayList<IDOMCandidate> cContainer = new ArrayList<IDOMCandidate>();
            cContainer.add(candidate);
            scoreboard.put(uri, cContainer);
        } else {
            final ArrayList<IDOMCandidate> cContainer = scoreboard.get(uri);
            if (cContainer != null) {
                cContainer.add(candidate);
                scoreboard.put(uri, cContainer);
            }
        }
    }

    public void add(String uri, List<IDOMCandidate> list) {
        if (scoreboard.get(uri) == null) {
            final ArrayList<IDOMCandidate> cContainer = new ArrayList<IDOMCandidate>();
            cContainer.addAll(list);
            scoreboard.put(uri, cContainer);
        } else {
            final ArrayList<IDOMCandidate> cContainer = scoreboard.get(uri);
            if (cContainer != null) {
                cContainer.addAll(list);
                scoreboard.put(uri, cContainer);
            }
        }
    }

    public void add(IRDFNode subject, IDOMElement element) {
        if (scoreboard.get(subject.toString()) == null) {
            final ArrayList<IDOMCandidate> cContainer = new ArrayList<IDOMCandidate>();
            cContainer.add(new DOMCandidate(element, subject));
            scoreboard.put(subject.toString(), cContainer);
        } else {
            ArrayList<IDOMCandidate> cContainer = scoreboard.get(subject.toString());
            if (cContainer != null) {
                cContainer.add(new DOMCandidate(element, subject));
                scoreboard.put(subject.toString(), cContainer);
            }
        }
    }

    public void add(IRDFNode subject, ArrayList<IDOMElement> list) {
        if (scoreboard.get(subject.toString()) == null) {
            final ArrayList<IDOMCandidate> cContainer = new ArrayList<IDOMCandidate>();
            for (IDOMElement element : list) {
                cContainer.add(new DOMCandidate(element, subject));
            }
            scoreboard.put(subject.toString(), cContainer);
        } else {
            ArrayList<IDOMCandidate> cContainer = scoreboard.get(subject.toString());
            for (IDOMElement element : list) {
                cContainer.add(new DOMCandidate(element, subject));
            }
            scoreboard.put(subject.toString(), cContainer);
        }
    }

    public ArrayList<IDOMCandidate> get(String uri) {
        return scoreboard.get(uri);
    }

    public void put(String uri, ArrayList<IDOMCandidate> list) {
        scoreboard.put(uri, list);
    }

    public void process(ArrayList<IRDFStatement> statements) {
        if (statements == null || statements.isEmpty()) {
            return;
        }
        for (final IRDFStatement stmt : statements) {
            final Predicate predicate = predicateFactory.create(stmt.getPredicate(), processor);
            if (predicate != null) {
                predicate.analyze(stmt, this);
            }
        }
        // TODO: analyze this staff to get a better interface between the predicates and the candidateManager
        //final IRDFResource subject = stmt.getSubject();
        //final Predicate predicateHandler = predicateFactory.create(stmt.getPredicate(), dom);
        //if (scoreboard.get(subject.toString()) != null) {
        //final ArrayList<IDOMCandidate> cContainer = scoreboard.get(subject.toString());
        //final Iterator<IDOMCandidate> iterator = cContainer.iterator();
        //while (iterator.hasNext()) {
        //final IDOMCandidate candidate = iterator.next();
        //predicate.analyze(stmt, candidate);
        //}
        //} else {
        //predicate.analyze(stmt);
        //}
    }

    /**
     * Determine the list of candidates that are scored by the classification
     * algorithm. So each candidate will can be evaluated by their score.
     *
     * @return a list of scored candidates
     */
    public ArrayList<IDOMCandidate> getBest() {
        final ArrayList<IDOMCandidate> scoredList = new ArrayList<IDOMCandidate>();
        for (String subject : scoreboard.keySet()) {
            //logger.debug("subject=" + subject + ",size=" + scoreboard.get(subject).size());
            for (IDOMCandidate candidate : scoreboard.get(subject)) {
                //logger.debug(candidate);
                if (candidate.hasMatchedFeatures()) {
                    logger.debug(candidate);
                    scoredList.add(candidate);
                }
            }
        }
        for (String subject : exactMatch.keySet()) {
            //logger.debug("subject=" + subject + ",size=" + scoreboard.get(subject).size());
            for (IDOMCandidate candidate : exactMatch.get(subject)) {
                //logger.debug(candidate);
                if (candidate.hasMatchedFeatures()) {
                    logger.debug(candidate);
                    scoredList.add(candidate);
                }
            }
        }
        return scoredList;
    }
}

File: src/wpsr/parsers/dom/DOMCandidateManager.ts
///<reference path="./../../factory/PredicateFactory.ts" />
///<reference path="./../rdf/predicates/Predicate.ts" />
///<reference path="./../rdf/query/IRDFNode.d.ts" />
///<reference path="./../rdf/query/IRDFStatement.d.ts" />

module wpsr.parsers.dom {
 export class DOMCandidateManager {

    private predicateFactory: PredicateFactory = new PredicateFactory();
    private processor: DOMProcessor;
    // get assoziativ array
    private scoreboard: HashMap<String, IDOMCandidate[]>;
    private exactMatch: HashMap<String, IDOMCandidate[]>;

    constructor(processor: DOMProcessor) {
      exactMatch = new HashMap<String, IDOMCandidate[]>();
      scoreboard = new HashMap<String, IDOMCandidate[]>();
      this.processor = processor;
    }

    public exactMatch(uri: String, candidate: IDOMCandidate): void {
      if (candidate == null) {
        return;
      }
      if (exactMatch.get(uri) == null) {
        cContainer: IDOMCandidate[];
        cContainer.add(candidate);
        exactMatch.put(uri, cContainer);
      } else {
        cContainer: IDOMCandidate[] = exactMatch.get(uri);
        if (cContainer != null) {
          cContainer.add(candidate);
          exactMatch.put(uri, cContainer);
        }
      }
    }

    public remove(uri: String, candidate: IDOMCandidate): void {
      if (candidate == null) {
        return;
      }
      if (scoreboard.get(uri) == null) {
        return;
      } else {
        scoreboard.get(uri).remove(candidate);
      }
    }

    public add(uri: String, candidate: IDOMCandidate): void {
      if (candidate == null) {
        return;
      }
      if (scoreboard.get(uri) == null) {
        cContainer: IDOMCandidate[];
        cContainer.add(candidate);
        scoreboard.put(uri, cContainer);
      } else {
        cContainer: IDOMCandidate[] = scoreboard.get(uri);
        if (cContainer != null) {
          cContainer.add(candidate);
          scoreboard.put(uri, cContainer);
        }
      }
    }

    public add(uri: String, list: IDOMCandidate[]): void {
      if (scoreboard.get(uri) == null) {
        cContainer: IDOMCandidate[];
        cContainer.addAll(list);
        scoreboard.put(uri, cContainer);
      } else {
        cContainer: IDOMCandidate[] = scoreboard.get(uri);
        if (cContainer != null) {
          cContainer.addAll(list);
          scoreboard.put(uri, cContainer);
        }
      }
    }

    public add(subject: IRDFNode, element: IDOMElement): void {
      if (scoreboard.get(subject.toString()) == null) {
        cContainer: IDOMCandidate[];
        cContainer.add(new DOMCandidate(element, subject));
        scoreboard.put(subject.toString(), cContainer);
      } else {
        cContainer: IDOMCandidate[] = scoreboard.get(subject.toString());
        if (cContainer != null) {
          cContainer.add(new DOMCandidate(element, subject));
          scoreboard.put(subject.toString(), cContainer);
        }
      }
    }

    public add(subject: IRDFNode, list: IDOMElement[]): void {
      if (scoreboard.get(subject.toString()) == null) {
        cContainer IDOMCandidate[];
        for (element: IDOMElement in list) {
          cContainer.add(new DOMCandidate(element, subject));
        }
        scoreboard.put(subject.toString(), cContainer);
      } else {
        cContainer: IDOMCandidate[] = scoreboard.get(subject.toString());
        for (element: IDOMElement in list) {
          cContainer.add(new DOMCandidate(element, subject));
        }
        scoreboard.put(subject.toString(), cContainer);
      }
    }

    public get(uri: String): IDOMCandidate[] {
      return scoreboard.get(uri);
    }

    public put(uri: String, list: IDOMCandidate[]): void {
      scoreboard.put(uri, list);
    }

    public process(statements: IRDFStatement[]): void {
      for (stmt: IRDFStatement in statements) {
        predicate: Predicate = predicateFactory.create(stmt.getPredicate());
        predicate.analyze(stmt, this);
      }
    }

    /**
     * Determine the list of candidates that are scored by the classification
     * algorithm. So each candidate will can be evaluated by their score.
     *
     * @return a list of scored candidates
     */
    public getBest(): IDOMCandidate[] {
      scoredList: IDOMCandidate[];
      for (subject: String in scoreboard.keySet()) {
        for (candidate: IDOMCandidate in scoreboard.get(subject)) {
          if (candidate.hasMatchedFeatures()) {
            scoredList.add(candidate);
          }
        }
      }
      for (subject: String in exactMatch.keySet()) {
        for (IDOMCandidate candidate : exactMatch.get(subject)) {
          if (candidate.hasMatchedFeatures()) {
            scoredList.add(candidate);
          }
        }
      }
      return scoredList;
    }
  }
}
